### ISSUE TITLE
Enable mocking static methods in Mockito

### ISSUE THREAD

**Issue by mockitoguy** [mockito/mockito#1013]
# Static methods mocking with Mockito

This is a placeholder ticket for enabling mocking static methods in Mockito. Example action items that are totally negotiable and can ran in parallel. We're looking for someone who can lead this effort.

- [x] Research + discuss whether it is a good idea to enable static methods mocking in Mockito. The theory is that it is useful for legacy code, which is most code in the world.
- [x] Research on how other mocking frameworks do that and whether it is considered useful feature for their users.
- [x] Design and present for discussion an API for static mocking (slightly relevant ticket: #643)
- [x] Work with @raphw / ByteBuddy to come up with hacky prototype (the hackier, the better!)
- [x] Mold the prototype with the API, remove enough rough edges so that the feature is good enough for incubating rollout
- [ ] SHIP IT!

**Comment by TimvdLippe** [comment]
I am torn on this one. If we provide this possibility, we might invite developers into bad practices. Every time a developer uses this feature, it should be very explicit that it is bad practice imo. Would it be possible to come up with a solution that is clear on this front, but still allows the developer to opt-in?

**Comment by szczepiq** [comment]
Great feedback. That's the first thing to figure out when working on this ticket :)

**Comment by raphw** [comment]
the way I saw mocking of static methods for myself was to only mock calls to static methods that happen within a mock. This would be useful, for example, for implementing spies in a situation like:

```java
class Foo {
  String bar() { return Util.makeString(); }
}
```
With Mockito, I would like to see something like:
```java
Foo foo = spy(new Foo());
doReturn("foo").when(foo).invokesStatic(Util.class).makeString();
assertThat(foo.bar()).isEqualTo("foo");
```

What do you think? As for a hack, I am missing a component in Byte Buddy which I was supposed to write for a customer at some point what did however not pull through. I am not currently in a position to spend so much time to build this component, due to my personal situation, but please prototype away. I think the API and spec work is crucial to make this a success.

**Comment by szczepiq** [comment]
Great feedback! Thank you Rafael for sharing your thoughts.

Is the component you mentioned a lot of work in ByteBuddy? Is this something that we can ask the community to help out or you're the best person to take on?

I'd rather see more robust API but your API idea is certainly workable. Mocking statics should be rare and our API should be optimized for common use cases and not edge/rare cases. We can always iterate based on feedback :)

Here's my preference as of today:

```java
//setup
mockStatic(Util.class);

//then
doAnswer().when(() -> Util.foo());
when(Util.foo()).thenReturn("foo");
verify(() -> Util.foo());
verify(never(), () -> Util.foo());
```

If we choose to support mocking static methods, I'd rather offer API that is robust enough to support all kinds of use cases. I like when tools are separated from policies and are opinionated but not dogmatic :) If we think that mocking static is disgusting, we can ship it in a different jar called "mockito-for-crappy-code", loosely following @TimvdLippe idea.

The main use cases for mocking statics I see are:
 - legacy code (I really, really want to write a unit test but I don't dare to change some ancient ugly code)
 - dealing with some awkward static 3rd party APIs. This scenario is currently workable by developing some injectable API layer on top of 3rd party statics. However, the workaround could be cumbersome and can spoil the clarity of codebase.

Without addressing above 2 use cases, developers around the world will be looking for help in tools like Powermockito, and such.

The biggest downside of offering static mocking is making it too easy to test crappy, procedural code full of static methods. We would remove a motivation to refactor the code into clean OO / DI.

**Comment by TimvdLippe** [comment]
Should we also send out a Tweet to gather some community insights?

**Comment by tmurakami** [comment]
If this feature uses the Java Instrumentation API, it will not work on Android devices, so it might confuse Android developers.
If so, I think it should not be integrated into the `mockito-core` artifact.

**Comment by bencampion** [comment]
I came across this issue by chance (I was just curious to know what you were planning for Mockito 3), but I'd like to share my experiences of working with static method mocking.

I used to work on several large codebases that made extensive use of static methods and used PowerMock (and PowerMockito) heavily for testing. My main memories of that time were that the tests were really slow to run (in some cases tests that use PowerMock were literally ten times slower than those that didn't) and that we ultimately had to remove all usages of PowerMock because it wasn't compatible with newer versions of Java. I also recall there being incompatibilities with some libraries were using that required special setup in the tests to make them pass, although I no longer remember the details of that.

I can understand concerns about promoting bad code, but as a Mockito user I'd be more worried about the impact on performance and robustness. Not all of the code we were using PowerMock for was legacy. Some of it was new code written by an inexperienced team (of which I was part of) and knowing that Mockito devs disapproved of our design patterns probably wouldn't have made any difference.

This was a few years ago now and techniques for mocking static methods may have improved significantly since then. If you think there's a performant and robust way to implement this feature then I'd welcome it (even though I'd hope I never have to use it).

**Comment by szczepiq** [comment]
Really good feedback, thank you guys. I helps us make good decisions about the features/API we plan.

**Comment by szpak** [comment]
In my opinion in general testing static methods is a bad idea and it would be better to refactor the code, however, sometimes it's something that is needed. For instance, testing a default method `given()` in [mockito-java8 interface](https://github.com/szpak/mockito-java8/blob/master/src/main/java/info/solidsoft/mockito/java8/api/WithBDDMockito.java#L68) delegating to a static method in `BDDMockito.given()` the easiest solution I see is to generate in runtime a list of static methods in `BDDMockito` and execute parameterized test for every of them verifying that a corresponding method in `WithBDDMockito` interface delegates to it with proper parameters. In that case static method mocking would be beneficial. Maybe I will wait for PowerMockito for Mockito v2 to test Mockito itself...

Alternatively I would need to play with AOP and load-time weaving, code generation in runtime or some other not very pleasant to use technique (unless you have some good ideas how to do it easier).

**Comment by szczepiq** [comment]
@szpak, thank you for the feedback! Interesting use case.

I think we all agree that mocking statics is a bad practice and an anti-pattern in general. The question is whether we enforce the "no mocking statics" policy (e.g. not offer the feature in the tool) or let the user decide to enforce it or not (e.g. offer the feature).

**Comment by Krazybug** [comment]
Please, consider the legacy code issue. Sometimes you don't have any choice. 
Enforcing or not a practice shouldn't be the decision of a framework but of the team.
Also, for now, we're using Powermock as a workaround for this need but it's not compliant with the last versions of Mockito. We're blocked with 1.9.x version of Mockito.
It's another good reason to get this feature: Limit dependencies.

We already need all these libraries to get decent unit test in pure Java world:
JUnit,
JUnitParams as Junit Parametrized are awfull
AssertJ for expressive and clear assertions
Mockito ... and Powermock

and i shouldn't mention other peripheral libraries:
DBUnit,
SpringDBunit
JBehave, JGiven for BDD style .... 

Then you can write all your testing stuf in Groovy and you only need ... Spock


Please! Let user choose. Othewise, why did you provide private method mocking in the recent versions athough it's considered as a "bad" practice for TDD purists ? 

Finally, thanks for your great work. Mockito is really a nice framework

**Comment by jlink** [comment]
Mocking static methods is a different use case, maybe that means it should be a different tool. 

Answer the following questions (to yourself):
- Do you want to support a double tool in the long term? 
- Does it use the same set of dependencies and only those? 
- Does mocking static methods fall nicely into the current design?

If you answer all 3 with "Yes", then go ahead. If at least one is a "No" consider the alternatives:
- Build a Mockito extension
- Build a new tool and call it Mackarita
- Leave the other tool to someone else

Just my two cents.

**Comment by karollewandowski** [comment]
I agree with most of you and think that mocking static methods is not a good idea.
If some method uses static method provided by legacy library/framework, then it can be easily wrapped in object and provided as mockable dependency. It's not perfect solution, but is educational - shows developers that writing too complex static util methods hurts and makes code hard to test. If Mockito had such feature, many of developers would consider it as acceptable practise and produce bad code.
Another issue is parallel tests execution. In my current project we used to use PowerMockito to mock static methods and after some time our tests started to fail because of concurrency issues:
1. Thread A mocked static method X.y and stopped.
2. Thread B mocked static method X.y and stopped.
3. Thread A was awaken and run test - it failed because thread B overwritten static method's behaviour expected in test run by A.
We ended up with wrapping static methods in injectable objects and removed PowerMockito from project dependencies.

**Comment by rdicroce** [comment]
I had asked about this at some point in the past and was told it was being considered. Glad to see that's actually happening.

In my opinion, support for mocking static methods is a good idea for the simple reason that the standard Java classes that ship with the JRE have tons of static methods. Consider the NIO.2 API as an example. If you want to read a file using that API, you might do something like:
```java
Files.readAllLines(Paths.get("myfile"));
```

Now, is there a way to test this without mocking static methods? Obviously yes; you could make a IFiles interface and then make a JREFilesImpl that passes the calls through to the static methods. But that's a bunch of extra effort, it adds complexity to your code, and obscures the methods that are actually being called. In my opinion, I shouldn't need to do that to test my code's ability to properly read some file.

So +1 for the ability to mock static methods.

**Comment by karollewandowski** [comment]
> Now, is there a way to test this without mocking static methods?

Well, in my opinion file name/path from your example is good candidate for passing it to method/setter/constructor instead of hardcoding it and you can just create test file in filesystem. Such Java API methods are considered as simple and reliable and there is no need to mock them like you wouldn't mock `java.util.List` if your method operated on data from given list.

**Comment by rdicroce** [comment]
There are at least 2 problems with that argument: a) you can't test behavior when exceptions occur, and b) you can't test behavior if the path is for a different OS. I realize the latter is an esoteric use case, but I actually am working on a project that is developed on Windows but runs exclusively on Linux.

**Comment by TimvdLippe** [comment]
I think the recurring theme right now is: either you are pro or you are strictly against it. Personally, I am for clean tests and therefore consider static mocking very bad practice. However, our users apparently have usecases that require (intrusive?) solutions such as PowerMock. Given that these users opt for such solutions signals that the other solution would be no tests at all, and that is probably what we would never want.

Given that solutions like PowerMock are built not only on Mockito, but also other libraries, they can not be easily updated like Mockito can. This incentives Mockito to solve static mocking, if users opt-in for it.

Even though I inherently disagree with static mocking, I think it is only fair that we as Mockito developers offer users the option to opt-in a less intrusive and upgrade-compatible solution. Therefore, I would vouch for a different artifact (probably called `mockito-legacy`), which offers static mocking. This should signal our users that this is for legacy purposes and should be strictly discouraged, but at least gives them a way out. This particular solution would tick off all points of @jlink

In the end, no matter which outcome, there will always be disappointed developers. We will not be able to find a golden solution for this particular problem and I am fairly convinced it will never happen either.

We can always try it out with the new artifact and discontinue it later if it is significantly misused.

One sidenote: putting everything in a new artifact would scatter the mockito ecosystem. We should probably find a better solution for this, potentially in Mockito 3.

**Comment by marcingrzejszczak** [comment]
I disagree with @TimvdLippe.  I don't believe that in an open source world you can "try it out with the new artifact and discontinue it later if it is significantly misused.". Once it's there in the library users will require it to be there. You'll have to support that feature etc. Either it's there or not. Not often can you easily deprecate sth and roll it back afterwards. Typically it requires a couple of release cycles.

@TimvdLippe you're mentioning this " However, our users apparently have usecases that require (intrusive?) solutions such as PowerMock". I have a comment on that. As an exercise from time to time I'm trying to answer PowerMock questions on StackOverflow. Typically someone says that there's a static method somewhere that they're calling and it does some logic that they want to stub. The question I often ask to those people is "Why are you doing this?". Would you really want to stub a call to `StringUtils` ? Another frequent problem is that the design is flawed and that's why one is looking for hacks. (e.g. http://stackoverflow.com/questions/37059406/how-can-i-get-powermock-to-return-the-expected-value-from-a-static-method/37066436#37066436 , http://stackoverflow.com/questions/37052069/program-termination-during-quartz-scheduler-verification-using-power-mockito/37066521#37066521 , http://stackoverflow.com/questions/31840964/powermock-private-method-with-null-pointer-exception/37066400#37066400 or http://stackoverflow.com/questions/32195928/getting-powermockito-to-mock-a-static-method-on-an-interface/32537392#32537392)

That's why I fully agree with @karollewandowski . "Every problem can be solved with a layer of abstraction" ;) You can't mock a static method? Ok then, wrap it in a class which you can stub. Is it wrong? Quite the contrary. Rarely should you call static methods directly and if you do, most likely these are utils that you don't want to stub. You should encapsulate the logic of a static method in an object that makes business sense to use it.

Summing it up giving people a tool to stub static methods is making them even easier to write bad code. Instead of thinking of how to fix the design they'll be able to make the design even worse.

**Comment by jlink** [comment]
@TimvdLippe Just one thing: If you create a new artifact like "mockito-legacy" communicate clearly if you're planning to support it mid/long-term or if it's just an experiment. Some people are willing to experiment together with you, others will be pissed off when you quit support after they have heavily used it.

**Comment by marcingrzejszczak** [comment]
@rdicroce I completely disagree with this statement:

> Now, is there a way to test this without mocking static methods? Obviously yes; you could make a IFiles interface and then make a JREFilesImpl that passes the calls through to the static methods. But that's a bunch of extra effort, it adds complexity to your code, and obscures the methods that are actually being called. In my opinion, I shouldn't need to do that to test my code's ability to properly read some file.

You're design is wrong. If you had 

```
Files.readAllLines(Paths.get(myFilePath));
```

You could via a constructor inject `myFilePath` to point to your test resources. You don't even need to create any additional classes.

**Comment by ChristianSchwarz** [comment]
The Android community would like mock static methods. 
Why? : The Android SDK provides many static utility methods with classes like TextUtils. The downside is that there implementation is only provided on the device or emulator. When users run Unit-Test on there local machine they will get the famous "Method not mocked" exception.

> The android.jar file that is used to run unit tests does not contain any actual code - that is provided by the Android system image on real devices. Instead, all methods throw exceptions (by default). This is to make sure your unit tests only test your code and do not depend on any particular behaviour of the Android platform (that you have not explicitly mocked e.g. using Mockito). 

They are many workarounds for this issue like PowerMock or Roboelectric. Roboelectric is a great testing framework but it is damn slow and maintainance intensiv. PowerMock is not stable state to be used with Mockito 2. So I think at least Android users will love this feature.

**Comment by dbacinski** [comment]
@ChristianSchwarz I do not agree, using Android static utils or Android framework in your business logic is a bad smell. Well structured Android app do not need mocking static methods, nor Roboelectric. Calls to Android framework must be abstracted and current Mockito implementation helps to enforce that. If you want to test view layer then go with Instrumentation tests and Espresso.

**Comment by szczepiq** [comment]
@dbacinski, I am not an expert on Android so bear with me, please :) Adding extra layer introduces more method calls which can be problematic on large apps that hit dex method limit, right?

**Comment by dbacinski** [comment]
@szczepiq you are right that you need additional methods to abstract Android apis, but this is a cost of good architecture. Dex method limit is not a huge problem anymore because now there is native platform support for multidex apps.

**Comment by rdicroce** [comment]
@marcingrzejszczak Your response a) does not address the call to readAllLines(), and b) does not address either of the issues I raised in https://github.com/mockito/mockito/issues/1013#issuecomment-291602703

To all: it seems to me there's a fundamental divide in this thread between people who are of the opinion that it's fine to just make a wrapper class for static calls, and people who are of the opinion that wrapper classes add bloat and shouldn't be necessary. I don't see either side convincing the other, so I'm not sure where that leaves us.

**Comment by marcingrzejszczak** [comment]
> Your response a) does not address the call to readAllLines(), and b) does not address either of the issues I raised in #1013 (comment)

@rdicroce I haven't explicitly but the answer is simple

> a) does not address the call to readAllLines(),

If the path is a parameter you don't have to mock the call at all. You can pass 

a) a path that exist - to test the positive scenario
b) a path that doesn't exist - that way it will blow up and you'll test an exception

> . I realize the latter is an esoteric use case, but I actually am working on a project that is developed on Windows but runs exclusively on Linux.

You can write a couple of tests that are exclusively executed depending on the OS (example for Windows - http://stackoverflow.com/questions/23410738/run-unit-tests-only-on-windows ). Again, if you use parameters instead of magic values then you can do basically whatever you want.

> To all: it seems to me there's a fundamental divide in this thread between people who are of the opinion that it's fine to just make a wrapper class for static calls, and people who are of the opinion that wrapper classes add bloat and shouldn't be necessary. I don't see either side convincing the other, so I'm not sure where that leaves us.

I'd say that the divide is between people who want to design their code properly and those who want to take the path of least resistance.

**Comment by pioorg** [comment]
>I'd say that the divide is between people who want to design their code properly and those who want to take the path of least resistance.

Static members aren't something good, I hope it's quite obvious. But Java has them and will support them, whether we like it or not. So I'd say that educating and influencing is good, forcing might be not. It's like "oooh, your code is using statics, so it can't be tested with Mockito".
(Not to mention that some people would like to have nice unclebobish tests but aren't allowed to refactor the code they have to maintain. I know it's sick, but hey, this is reality sometimes.)

**Comment by rdicroce** [comment]
@marcingrzejszczak 

> If the path is a parameter you don't have to mock the call at all. You can pass
>
> a) a path that exists - to test the positive scenario
> b) a path that doesn't exist - that way it will blow up and you'll test an exception

That may solve this particular case. What about other static APIs that can throw a variety of different exceptions? Are you saying users should go through the hassle of setting up every possible scenario? What happens if there are scenarios the user can't set up? You may as well not bother with a mocking framework at all in that case.

> You can write a couple of tests that are exclusively executed depending on the OS (example for Windows - http://stackoverflow.com/questions/23410738/run-unit-tests-only-on-windows ). Again, if you use parameters instead of magic values then you can do basically whatever you want.

This makes it impossible to run the tests in the environment where the code is developed. Again, I know this is an esoteric use case, but it is a valid one IMO.

> I'd say that the divide is between people who want to design their code properly and those who want to take the path of least resistance.

I won't speak for anyone else in favor of mocking static methods, but I don't see that as designing code improperly. Adding a whole extra layer of abstraction that's completely unnecessary outside of testing? That does strike me as being improper.

**Comment by marcingrzejszczak** [comment]
This will be my last answer cause obviously we are not going anywhere with this.

@pioorg

> Static members aren't something good, I hope it's quite obvious. But Java has them and will support them, whether we like it or not...

Yes it's obvious. Someone has to afterwards maintain that code. Both the business and the library. Another thing is that you don't have to mock all of those calls. Why would you stub a `StringUtils` class ? Makes absolutely no sense. 

@rdicroce

>  Are you saying users should go through the hassle of setting up every possible scenario? What happens if there are scenarios the user can't set up...

Of course not. You can catch a generic exception and that's enough. 

> This makes it impossible to run the tests in the environment where the code is developed

That's false. Some of your tests will be executed only on Windows. Some others on Linux etc. 

> I won't speak for anyone else in favor of mocking static methods, but I don't see that as designing code improperly

That's why I would just call the static method. If in your static method you have a lot of complex logic then it should get wrapped.

All in all - I think that this is a wrong decision for Mockito. But that's only my opinion.

**Comment by pioorg** [comment]
> Yes it's obvious. Someone has to afterwards maintain that code. Both the business and the library. Another thing is that you don't have to mock all of those calls. Why would you stub a StringUtils class ? Makes absolutely no sense.

With all the respect, I don't care about any *Utils classes or any other particular code snippets.
What I mean is that:

1. statics are valid from syntax PoV
2. they already exist in legacy code 
3. some folks can't change that code "only because we'd like to have Mockito in tests"

So if one can't change the code, refactor it and add the wrappers around the static calls, they also can't use Mockito because "Mockito doesn't support statics". To me it comes to "bad code -> bad tests". Sorry ;-)
While I perfectly understand there are reasons not to include it in "standard" distribution artefact, having another one being mockito-extension-for-ppl-who-dont-have-nice-code.jar acceptable.

**Comment by ChristianSchwarz** [comment]
Just some toughts about  the prevailing opinion "statics methods are evil". 
From my experience there are 2 different kinds of static methods "good" and "evil", i wouldn't say that all static methods are evil.

 * Static methods are "evil"  when they alter the state of a of a static member or in other words they depend on a state. When such a method is called the result may be an other when called in the future with the same input.
E.g:
```java
 class LogManager {
     static LoggerPlugin plugin = ...
     static void logInfo(String message){
        plugin.write(INFO,message);
     }
 }
```
 When the LoggerPlugin changes during runtime the result may be different, e.g logged to a different media.

 *  Static methods are also "evil" when they call a static method that is "evil" as described above, cause they inherit the nasty side effects.
E.g:
```java
 class Log{
  static void info(String message){
      LoggerMananger.logInfo(message); 
  }
 }
```


 * Static methods are are "good" when they don't depend on a state / access a static member and thus have the same result on every call.
E.g:  `Math.max(int,int) `


In the java eco system exists both evil and good static methods. I don't see a much points to mock "good" static methods. It makes only sense when they do quite complex stuff and you don't want to mock all bits of there input to satisfy there implementation. Even in famous frameworks we are faced with "evil" static methods (hello `Singleton.getInstance()`). To handle these evil cases mockito can do it's part and support the poor devs among us who have to deal with it. I wouldn't blame mockito if devs start to introduce the "evil's"  to there code base. Mockito is a mocking framework, not a good style enforcement tool, thats the task for the team and/or lead developer.

**Comment by pioorg** [comment]
@ChristianSchwarz 
> Static methods are are "good" when they don't depend on a state / access a static member and thus have the same result on every call.
True, there are even some schools in which all private methods should be static, don't depend on anything apart arguments, don't change the state of any object and don't do anything beside computing the result. So they're functions, just like sin(angle) is: the angle isn't changed, and sin is only the result of the angle, always the same (provided the geometry is always the same ;-) )
I find this functional style of programming very good, but in fact it has rather not much to do with statics as such. IMHO sometimes it's better to tell newbies "don't use statics, 'couse they're generally bad"... It's a kind of "lie for children".

**Comment by kkapelon** [comment]
Unfortunately as history has shown multiple times, when developers see a tool and its use cases, they do not always understand the impact and weight of its features. They consider all features equal and do not understand that some of them should only be used as a last resort.

The canonical example in the Java world is the abuse of Java introspection/reflection and AOP for even trivial scenarios. Even outside the Java world we see the misuse of advanced features by developers who use them "simply because they are there". 

If we are to go forward with this (and I am strongly against this, as I considered a major advantage of Mockito the fact that it did not allow testing of static methods) I would suggest the following:

- Create a "mockito-legacy" artifact/extension/add-on 
- Require a configuration file option for Mockito to actually enable this add-on
- Print in the log file when the test runs a huge warning (similar to what Maven does when you don't define a version) that using this add-on is an anti-pattern and a code-smell
- Provide a link to the user that has a page with common refactoring scenarios that show how we can get rid of static methods.

Maybe I am missing something, but I really would like some examples on how legacy code cannot be refactored to remove the static dependency. Unless of course you cannot change the code at all, which for me makes the creation of unit tests questionable (because if a unit test fails what are you going to do then?)

**Comment by pioorg** [comment]
@kkapelon: I also wish all Java code was pure OO design and that everyone actually knows what that means ;-) I hope such a huge warning "you're doing it wrong" could educate and maybe convert a few folks.

> Maybe I am missing something, but I really would like some examples on how legacy code cannot be refactored to remove the static dependency.

It's not about the code itself. I've seen big name companies enforcing policies in which before bugfixing, the case has to be replicated in tests. (Kinda TDD, they claim sometimes.) Then, once the test fails just like ticket describes, the support team is allowed to change the code to make it pass (and sadly, usually minimum changes). It's rather difficult to make a test to cover statics if your testing tool set doesn't allow that.
It's like all these The Daily WTF stories in which "don't change it, it ain't broke, 'couse it works!", so the code is like Java 1.4...

**Comment by lukaseder** [comment]
This would be great! We could finally stop bikeshedding "static is *eeevil*" because not testable.

**Comment by pioorg** [comment]
> We could finally stop bikeshedding "static is eeevil" because not testable.

Well, IMHO static is eeevil in many cases not only because not testable :-D

**Comment by helmbold** [comment]
The mocking framework shouldn't restrict the design options - it's just not the job of a mocking framework! I've seen a lot of code polluted with mocking-only interfaces. And there are a lot of cases where a simple static method is a perfect design decision. `final` is also rarely used, because mocking frameworks (besides the great [JMockit](http://jmockit.org/)) have trouble with it.

**Comment by trumpetx** [comment]
DONT DO IT!  PowerMockito exists for those cases where you absolutely need to mock static methods.  Right now Mockito conforms to the commonly held opinion that static is for methods that are testable without mocks (think: StringUtils).  Static for other stuff should be discouraged (there are exceptions, obviously) - in those situations, encapsulating the static calls in a protected @VisibleForTesting method should suffice for unit testing and for those other "legacy or not" situations, there's always PowerMockito.

**Comment by TimvdLippe** [comment]
@trumpetx I just wanted to note that the reason we are investigating this change is that some companies are in the legacy code situation, where adopting PowerMockito would be too disruptive. Therefore, these companies would be unable to test their classes. Even if they could adopt PowerMockito, the tests become significantly more difficult to understand, just to be able to test their legacy code.

The question is: should we as Mockito give the opt-in for these situations, as other options are too invasive? I think every Mockito core developer agrees that static mocking is bad and should not be done, but sadly we can not dictate what the industry does or has to deal with.

Just wanted to give perspective on why we are opening this issue.

**Comment by TimvdLippe** [comment]
Also for reference, I cross-posted this issue on Reddit and there some other arguments/viewpoints posted there: https://www.reddit.com/r/java/comments/64pdk1/mockito_discussion_on_introduction_of_mocking/

**Comment by pioorg** [comment]
@TimvdLippe 
Thanks for your link. The recent posts have led me to another...
http://mvnrepository.com/artifact/org.powermock/powermock-api-mockito/usages
I'm just wondering why so many of them... And I didn't consider them so far as legacy ;-)

**Comment by AnEmortalKid** [comment]
I haven't had experience with ByteBuddy but I'd be willing to give it a shot. As far as names, how do we feel about `mockito-gore` to really indicate you need a strong reason to use this jar?

**Comment by pioorg** [comment]
Just a few more words from me, maybe last ones ;-)
There's a method in Mockito:
`org.mockito.Matchers#isNull(java.lang.Class<T>)`
Isn't NULL considered to be "The Billion Dollar Mistake"? Shouldn't the programmers all around the world be educated that passing a null reference (instead of a `java.util.Optional`) is a very bad practice?
Maybe this `isNull` method should be deprecated and eventually removed in future versions?

I've seen a lot of code being polluted by `if (reference == null)` and `catch (NullPointerException e)`... More than by "incorrect use of statics". (But YMMV.)
If supporting statics is going to create such a bad habits among programmers AND Mockito is meant to be a forced education, then I guess support for NULLs should be removed in future versions and people should write wrappers to handle NULLs from their legacy code, as NULL is worse habit IMHO. ;-)

Unless Mockito is meant to be a testing/mocking framework, not forced education course.
Thank you.

**Comment by filiphr** [comment]
I would like to add some comments from my own experience. I completely understand the people that are saying to add an abstraction to the static methods that you are invoking. However, in a legacy code that is expensive and requires a lot of effort. 

Additionally, there are frameworks like [GWT](http://www.gwtproject.org/) that support polymorphism, but not like you would think. One needs to call a static method that will do the creation of the object on runtime and runtime there is Javascript. There ware ways that you can write tests that will emulated the client code for GWT, but they are really slow. Currently, we use PowerMockito with Mockito 1.9.5, because we can't update to a later Mockito. However, if Mockito had this support it would be much easier to do the testing.

**Comment by TimvdLippe** [comment]
This discussion has been very active and helpful, thank you everyone for participating!

I think we have sufficient background information to make a final decision @mockito/core Shall we discuss this on Slack and write a wiki page on our full analysis and final verdict? Then we can inform the community for a final review (of a potential implementation) and round up this feature.

**Comment by mattnelson** [comment]
> Shall we discuss this on Slack and write a wiki page on our full analysis and final verdict? Then we can inform the community for a final review (of a potential implementation) and round up this feature.

Has the verdict on static support been reached?

**Comment by TimvdLippe** [comment]
We have not. Every single core member has been swamped with various IRL activities. Probably this fall we will be back in full force. This is one of the features that we will be addressing :smile:

**Comment by szczepiq** [comment]
The team was not actively working on this. The ticket is marked "please contribute" for a reason ;) Currently I work with @thekingnothing on creating Mockito APIs so that Powermock can cleanly integrate (#1110) this will effectively provide reasonable support for static mocking. 

Hope this clarifies the status! Thank you for reminder - this is useful for us to manage our priorities.

**Comment by ogregoire** [comment]
Static methods are totally OK. You just have to know when to use them.

I think the following code is entirely correct and sufficiently abstracted for the purpose of a modern-day application:

    // guava imports
    class ApplicationResources {
      // points to a file in the .war
      private static final CharSource DEFAULT_CONFIG_SOURCE = Resources.asCharSource(Resources.get("etc/default.config"), UTF_8);
      static CharSource defaultConfigSource() {
        return DEFAULT_CONFIG_SOURCE;
      }
    }

I need to access a static value in my application. But for the purpose of proper testing, I need to test various default values, so it makes sense to provide another `CharSource` during the tests than at runtime.

Just imagine that this is embedded in some JEE application. There is no sense in writing this method as non-static. I could write `new ApplicationResources().defaultConfigSource()`. What is that code? Is it real? Others might imagine I make this class a (singleton) bean. A bean requires the class to be public. I don't want any of its method to be accessed outside of the package, so there are no reasons to make this class public, so this class has no reasons to be a bean.

However I still have to test classes that use it. But I'm stuck in the testing because I just can't mock that method (besides using all the workarounds mentioned above).

I agree that mocking static method should be used very, very rarely, but it should exist for those rare cases.

**Comment by raphw** [comment]
Personally, I plan to add this feature. I still need to do some ground work in Byte Buddy to make this feasible. It is not as easy as I hoped, unfortunately and it might take some time. The same goes for the seamless integration with PowerMock.

If anybody would be willing to sponsor the development, I am happy to use paid hours, within my open source commitment, this might take a few more month to complete but I am certain to finish this at some point. Java 9 compatibility takes priority at the moment.

**Comment by moltmann** [comment]
**tl;dr:** mocking static methods can be implemented mostly like regular mocking with [a few quirks when spying on static methods with side-effects](https://github.com/moltmann/dexmaker/wiki/Experiments-with-mocking-static-methods-via-mockito#spying-and-doreturn).

I implemented a [prototype](https://github.com/linkedin/dexmaker/compare/master...moltmann:staticMock) of this for the dexmaker based mocking used on Android. Examples and tests can be found in [MockStatic.java in the diff](https://github.com/linkedin/dexmaker/compare/master...moltmann:staticMock#diff-923767ff4d088765b531f9151e31b1c3R127).

I summarized my observations on a [wiki page](https://github.com/moltmann/dexmaker/wiki/Experiments-with-mocking-static-methods-via-mockito).

**Comment by moltmann** [comment]
I think I found a [nicer version for mocking static methods via doReturn().when()](https://github.com/moltmann/dexmaker/wiki/Experiments-with-mocking-static-methods-via-mockito#create-temporary-object-just-for-calling-static-methods). So the total interface for static methods could be designed as:

    try (StaticMockingInProgress<MyClass> token = spyStatic(MyClass.class)) {
        when(MyClass.aStaticMethod()).thenReturn(“mockedValue”);
        assertEquals(“mockedValue”, MyClass.aStaticMethod());
        
        doReturn(“fakeValue”).whenStatic(MyClass.class).anotherStaticMethod();
        assertEquals(“fakeValue”, MyClass.anotherStaticMethod());
    }
    
    assertEquals(“originalValue”, MyClass.aStaticMethod());
    assertEquals(“originalValue”, MyClass.anotherStaticMethod());

In Summary we could limit the necessary changed to:
- mockStatic/spyStatic instead of mock/spy. This is required as mockStatic/spyStatic needs to return a token to define the scope of the mocking instead of a mocked object
- doReturn(...).whenStatic(Class) instead of doReturn(...).when(Object). This is necessary as when mocking a static method, there is no mocked object

**Comment by moltmann** [comment]
Addendum to previous comment. I think we should also have a `resetStatic(Class<T> class)` method. Then using the StaticMockingInProgress object becomes optional but recommended.

**Comment by mockitoguy** [comment]
Hey @moltmann, thank you for taking a stab at the prototype!

Given that we provide an API for mocking statics in mockito-core, will you be able to come up with the implementation in dexmaker? 

Can you implement static mocking in dexmaker right now, with current version of mockito-core API? For example, can you add ```MockitoStatic``` class to dexmaker and offer static mocking to dexmaker users? If not, what are the problems with mockito-core API that block you.

**Comment by moltmann** [comment]
I am trying to sum up all discussion I had and what I learnt from this thread.

I updated my prototype at https://github.com/moltmann/dexmaker/tree/staticMock using 
- whenStatic 
- explicit finishStaticMocking

# Stubbing

How does "when" work for static methods?

## when(λambda)

### Suggested APIs

```
when(() -> MyClass.myMethod()).thenReturn("mocked return value");

doReturn("mocked return value").when(() -> MyClass.myMethod());

verify(() -> MyClass.myMethod());
```

### Problems

#### Not clear what method is called

If the lambda calls two methods, e.g.:

```
doReturn("mocked return value").when(() -> MyClass.firstMethod().secondMethod());
```

or 

```
doReturn("mocked return value").when(() -> {
    Object o = MyClass.firstMethod();
    return o.secondMethod()
});
```

Which method should we set up mocking for? `MyClass#firstMethod()` or `?#secondMethod()`.

- We need to make sure the return type matches the parameter of doReturn. We only know the return type of the the lambda, i.e. the return value of secondMethod. Hence we have to set up mocking for the second method.
  We don't set up mocking for `MyClass.firstMethod()`. What should `MyClass.firstMethod()` do? Call the real method? Return null? To make the code above work `MyClass.firstMethod()` has to call the real method. But this might have side effects. The point of the `doReturn(...).when(...)` syntax is to avoid side-effects.
- The second method is most likely an instance (non static) method. Should we allow setup for instance methods using the when-with-lambda?

I think we need to enforce only a single mocked method call on a mock in the lambda. Hence we have to listen to all possible methods calls, and intercept them. Once a second one or one of the wrong return type is intercepted we need to throw an exception. Hence above examples will throw an exception.

##### Possible extension

One possible extension would be do declare a restriction on what to intercept like

```
doReturn("mocked return value").when(MyOtherClass.class, () -> {
    MyOtherClass o = MyClass.firstMethod();
    return o.secondMethod()
});
```

Which declares that we want to set up mocking for `MyOtherClass`, not `MyClass`.

#### Conflict between old and new doReturn-when

```
doReturn(() -> "A").when(() -> MyClass.firstMethod());
```

In above case Mockito needs to decide if this is the new setup method or an incomplete version of 

```
Supplier<String> m = mock(Supplier.class);

doReturn(() -> "A").when(m).get();
```

it can do this by checking if the parameter to m is a mock during runtime, but the when(λambda) API needs to support both cases, hence when(λambda) has to return a lambda type, i.e.

```
LambdaStubber<Supplier<T>> doReturn(Supplier<T> ret) {
    return new LambdaStubber<Supplier<T>>(ret);
}

public static class LambdaStubber<T> implements Stubber {
    Supplier<T> when(Supplier<T>) {
        ....
    }
}
```

##### Possible extension

This issue might be mitigated by introducing

```
doReturn(() -> "A").whenStatic(() -> MyClass.firstMethod());
```

## `T whenStatic(Class<T>)`

```
doReturn("mocked return value").whenStatic(MyClass.class).myStaticMethod();

verifyStatic(MyClass.class).myStaticMethod();
```

`doReturn(...).whenStatic(MyClass.class)` returns a mock instance of `MyClass` with all instance methods throwing a stubbing exception. So the user will not accidentally call a instance method on this special stubbing-mock.

### Problems

`doReturn(...).whenStatic(MyClass.class)` returns a mock instance of `MyClass`. Hence calling `doReturn(...).whenStatic(MyClass.class).myStaticMethod()` is equivalent to

```
MyClass m = mock(Myclass.class);
m.myStaticMethod();
```

While it is valid Java to call static methods on instances, all IDE code checkers, linters and tools like errorprone will light up like a christmas tree and tell you not to do it.

# Life-time of static mocking

Regular instance mocking is tied to an object's lifetime. Static methods do not have such a life-time, hence when to know when to reset the methods and stop collecting invocations.

## `mockStatic(Class)` -> `finishStaticMocking(Class)`

This is the most basic solution. The advantage is that this works with every possible test-framework and if someone wants to have a nicer API they can e.g. build a `AutoClosable` that automatically calls `finishStaticMocking(...)` when a try-with-resources block is exited.

## MockitoSessions

As far as I understood there can only be one session at a time. Hence scenarios with interleaved life-cycles of static mocking can not be supported. Example:

```
mockStatic(A.class)
mockStatic(B.class)

when(A.m()).thenReturn("mockA")
when(B.m()).thenReturn("mockB")

assertEquals("mockA", A.m());

finishStaticMocking(A.class)

assertEquals("mockB", B.m());

finishStaticMocking(A.class)
```

Of course if the user wants to use sessions (or other session like tools, such as the Mockito JUnit runner) they should apply to static mocking too.

**Comment by moltmann** [comment]
I implemented the recommendation of https://github.com/mockito/mockito/issues/1013#issuecomment-291045005 in :
- Code: https://github.com/moltmann/dexmaker/tree/staticMock
- JavaDoc: https://moltmann.github.io/dexmaker/com/android/dx/mockito/inline/extended/package-summary.html

It mostly works even though some parts of the implementation use nasty reflection.

**Comment by avanathan** [comment]
@moltmann Could you please open PR for your changes?

**Comment by moltmann** [comment]
@avanathan My code is currently based on [dexmaker-mockito](https://github.com/linkedin/dexmaker) and does not work with the bytebuddy mock maker used by vanilla Mockito. I think we should first agree on the interface before trying to make this work with bytebuddy.

**Comment by mockitoguy** [comment]
@moltmann, thank you for working on this! I plan to re-engage soon.

**Comment by binkley** [comment]
I didn't see it discussed, but the main use case for me wanting to mock a static field is logging.   Popular logging frameworks (especially SLF4J) store a log object in a static field `private static final Logger logger = LoggerFactory.getLogger(MyClass.class)`.

In tests I would like to mock this field, so I can validate logging calls at critical places in the business logic, for business reasons of audit trails and operations support.

**Comment by marcingrzejszczak** [comment]
Imo you shouldn't mock the logger but assert whether the logging output contains the desired value

**Comment by pioorg** [comment]
@binkley Can we go one step back?
If you have business reasons to write some audit logs, they _might_ be part of your business. Ergo: your model and your domain. That's perfectly fine in banks, insurance, security, other areas.
Are you really sure _techy_ logs are the way to satisfy the business needs?
E.g. in SLF4J/Logback it's a matter of the configuration for the precise runtime in production. So even if you test static calls, there's no guarantee that this will happen in real business. Someone may misconfigure SLF4J. So again, _maybe_ the audit should become explicit part of your model?

**Comment by ghsatpute** [comment]
There are some use cases where I've to mock the third party code while running unit tests. 

`IOUtils.toString(File file, ...)`
or 
`RequestContext.getCurrentContext()`

I'm not sure how and why should I refactor this code.

**Comment by GregDThomas** [comment]
@ghsatpute - The standard way is with a wrapper; e.g. 

```
public interface IOUtilsWrapper {
    default String toString(File file, ....) {
        return IOUtils.toString(File file, ...);
    }
}
```

And pass in either a mock of IOUtilsWrapper, or the real thing.

**Comment by ghsatpute** [comment]
@GregDThomas  
Are you suggesting that in my actual code I should use `IOUtilsWrapper` class? That would overkill if I have too many such methods? Or am I missing something?

**Comment by GregDThomas** [comment]
@ghsatpute  - in the absence of the ability to mock a static, yes. Alternatively, consider the user of PowerMock - https://github.com/powermock/powermock/wiki/Mockito#mocking-static-method

**Comment by iirekm** [comment]
I already wrote an extension on top of Mockito+AspectJ, with syntax like
```
when(() -> YourClass.staticMethod(...)).thenReturn(...)
```
It's available here: https://github.com/iirekm/varia/tree/develop/ajmock

Maybe the good way of thinking is:
- pure Mockito - for new, pretty, shiny, testable code
- AjMock - for legacy, untestable code to be able to add tests to it without modifying the source code (which is hard and may break something)

**Comment by CharlieReitzel** [comment]
With all of the text expended on the horrors of static methods, I have yet to read any solid justification for these strongly held opinions.   It really appears, to me, to be pure dogma and not engineering.

As a software developer, static methods have a great appeal.  The explicitly disavow any state.  That's a big deal and a very, very good thing.   Fwiw, in part due to their lack of state, static methods are themselves _more_ easily covered in JUnit tests.

The _only_ down sides to static methods, that I can see, are a) they cannot be overridden by derived classes, which leads to b) they are more difficult to mock in unit tests.

Downside a) is actually a good thing and easily worked around using a non-static wrapper (as illustrated above) in use cases that require such.

Which leaves mocking.   And only mocking.

So, if there is a good way to help deliver well tested code, I'm all for it.  So, please, add support for mocking static calls directly to Mockito core as soon as is convenient.

**Comment by dbacinski** [comment]
"The explicitly disavow any state."
so what's the point to mock/stub them, if it is just a pure logic.
Static methods actually allow state modification of any accessible part of app, that's the main problem why ppl would like to override its behaviour.

**Comment by anthofo** [comment]
From my point of view Mockito is a mocking tool, not a linter, it should not tell you whether your code is good or bad, or dictate how to write it. It's all a matter of taste, like the endless "tabs vs spaces" debate and you won't find 2 developpers with the same coding style, so just let people setting up their linting tool the way they want. And even if you think that static is evil, since static methods are allowed in Java anyway, your software can possibly use 3rd party librairies that use them and that you would like to mock, and wrapping everything is overkill (but that's also a matter of taste :) )

**Comment by GregDThomas** [comment]
I'd argue that a static method that uses the network or disk isn't a good static method - it relies too much on external inputs for it to be a pure function. But, that's personal preference, clearly many people do this, and as such I do think Mockito should offer the option to mock statics.

**Comment by avanathan** [comment]
> I am torn on this one. If we provide this possibility, we might invite developers into bad practices. Every time a developer uses this feature, it should be very explicit that it is bad practice imo. Would it be possible to come up with a solution that is clear on this front, but still allows the developer to opt-in?

IMHO, mockito should not be used as an enforcer for not using static methods. If we provide this feature in mockito - we could use mockito to test legacy code.

**Comment by CharlieReitzel** [comment]
>> "The explicitly disavow any state."
> so what's the point to mock/stub them, if it is just a pure logic.
> Static methods actually allow state modification of any accessible part of app, that's the main problem why ppl would like to override its behaviour.

In my case, we have a directory structure that contains hundreds of thousands of files.  These files were created by a complex, analytical application executed repeatedly over several years.  Different versions of that software wrote different flavors of the files in varying sub-directory structures.  I have a static class that builds paths to those files based on a small number of inputs.  All of the "state", which is based on known behavior of different versions of the analytic software, is known at compile time and has almost no dependencies.

To unit test software the consumes these files, without re-creating the runtime environment, it's really helpful to mock the results of the path generating static methods to point to files packaged as test resources (i.e. not accessible via any file system).

Again, the _only_ case against static methods is that Mockito doesn't support mocking them (or is preventing PowerMock from doing so)!!   Every other argument against boils down to a style preference.

Otherwise, there are any number of valid use cases where all state is known at compile time, but that it is very helpful to mock the results for unit test purposes.

**Comment by ghsatpute** [comment]
> I'd argue that a static method that uses the network or disk isn't a good static method - it relies too much on external inputs for it to be a pure function. But, that's personal preference, clearly many people do this, and as such I do think Mockito should offer the option to mock statics.

I think, people use static methods because they don't have any states. Sure there are too many external inputs, but there is _**no state**_. If I read the file a thousand times, I'll get the same output. And I don't think that if we make it no-static those _external inputs_ automatically gets solved. 

For stateful code, use class methods and for stateless methods use static.

**Comment by IAmHopp** [comment]
> > To all: it seems to me there's a fundamental divide in this thread between people who are of the opinion that it's fine to just make a wrapper class for static calls, and people who are of the opinion that wrapper classes add bloat and shouldn't be necessary. I don't see either side convincing the other, so I'm not sure where that leaves us.
> 
> I'd say that the divide is between people who want to design their code properly and those who want to take the path of least resistance (which isn't a wrong choice sometimes).

This really highlights the difference between discussing the needs of the feature and judging the development choices of other teams.

I think @rdicroce got it right. What we see in this issue is that many people are perfectly comfortable buidling wrapper classes for this single, exclusive need [or even PowerMock, which does a bunch stuff, but again for this single purpose] and others wish they could just test the code they already got.

For the first group of people, no action is needed, but also nothing would be lost -- They'd just keep using Mockito while skipping the static mocking. For the second group, I think Mockito [core, by the way] is lacking a feature that could genuinely help them.

In any case, thanks to the team for even investigating this. I hope it gets implemented.

**Comment by lfarkas** [comment]
what is the current state of this FR? is there any progress is there any decision?

**Comment by raphw** [comment]
This is not currently pursued. It is technically possible but it is a bigger venture and I am currently lacking the time for it.

**Comment by IonelLupu** [comment]
So, any plans on supporting this?

**Comment by ChristianSchwarz** [comment]
As @raphw said it is technically possible. I managed to write [a prototype to stub static methods](https://github.com/ChristianSchwarz/mockito-lambda/commit/a0101976f7af5db5effa39269ec69538eb6f2b78), it also allows to stub `static void` methods using a lambda like that:

```
Mockito.mockStatic(StaticUtil.class);
when(()->StaticUtil.voidFoo()).thenThrow(new RuntimeException());

// or  equivalent 
when(StaticUtil::voidFoo).thenThrow(new RuntimeException());
```

Unfortunatly the idea to use lambda as `when`  argument was rejected years ago (#828 ). 

Therefore the first step should be to define a scope and an API that needs to be accept by the mockito-core developers.

**Comment by jacky1193610322** [comment]
The issue exists for about 3 years, So any plan to support it?

**Comment by ChristianSchwarz** [comment]
> So any plan to support it?

The OP of this issue is the founder of mockito, so the answer is yes. Due to the lack of time non of the core developers can't work on it.  
I would like to work on it but I am afraid that a PR won't get reviewed or merged due to the shortage of time of the core mockito developers. An other option would be to fork mockito, but that undertaking is to much effort for me alone since you need to setup the whole release infrastructure.

Nevertheless everyone reading this is invited to participate, either by API discussion, prototypes or other toppics to drive this.

**Comment by TimvdLippe** [comment]
Getting this feature right takes a lot of effort and time, as Rafael pointed out. Moreover, in the current situation with the pandemic, we are finding ourselves with less time on our hands and are dealing with problems not related to programming 😢 

There is consensus in the core team that we do want this feature. We have written prototypes and explored this space. Based on these explorations, we had to conclude that it was more challenging than we initially expected. This is not something you can implement in 1 PR and has implications on the larger architecture of Mockito.

While you of course have the option to fork, these problems would remain. We expect you will have a difficult time implementing proper static method support in the current project structure. That's to say: it is probably possible to hack it in, but that is not good for longterm maintenance of the project. Given that Mockito is a mature project used by a lot of users, implementing these hacks would not be the appropriate decision for the large majority of our users.

In my personal judgement, this feature points to the issue that Mockito has matured over time, but we are now finding shortcomings in the current architecture. A better look at the internal implementation and a potential overhaul would help us plan support for this feature. However, I think we are obligated to maintain API stability and prevent a Mockito 2 upgrade scenario, where customers face a large cost of upgrading. Getting that balance right is difficult and time-consuming. Time that we do not have at this moment in time, given the state of the world and the effect that has on our mental/physical health.

**Comment by CharlieReitzel** [comment]
What is the state of Mockito + PowerMock?   Mostly, I use Mockito.  I end up using different versions on different projects.  My recollection is that, with newer versions of Mockito, it _intentionally_ no longer plays nice with PowerMock.   I.e. all that is required is to re-enable an API or two to re-nable PowerMock.  No need to duplicate that feature set.

**Comment by TimvdLippe** [comment]
I am not sure where you got that information from, but that is not our intention. If Powermock 2 and Mockito 3 don't work together, that is a bug.

**Comment by ghsatpute** [comment]
@CharlieReitzel I personally don't like using PowerMockito since Coverage tools show zero coverage for the classes in which we've used PowerMockito.

**Comment by CharlieReitzel** [comment]
@ghsatpute Hi Ganesh, I have heard that complaint in the past.  But have had a problem getting coverage measurements for the class under test while using PowerMock.    I am assuming, of course, you don't apply PowerMock to the class under test.

**Comment by CharlieReitzel** [comment]
@TimvdLippe Forgive me if I have lost the specifics.  It's been a couple years.  Last I checked, I could not use any version of PowerMock with any 2.x version of Mockito and the reason was that certain methods that were public in 1.x became private in 2.x.   If that issue has been resolved in the meantime, I would suggest the following:

1. Verify it is, in fact, possible mock static methods using PowerMock 2 and Mockito 3.
2. Close this issue as "Won't fix" with the guidance to use PowerMock for that purpose.

Personally, I agree that, if you find yourself wanting to mock a static method, first think hard about it and see if there isn't a better way to get the job done.   But, sometimes, you just need a sledgehammer!  :--)

**Comment by TimvdLippe** [comment]
@CharlieReitzel https://mvnrepository.com/artifact/org.powermock/powermock-api-mockito2 should be compatible with Mockito 2/3. For more info see https://github.com/powermock/powermock/wiki/Mockito-2-Maven

This ticket is about supporting static method mocking in Mockito, without having to rely on Powermock. Since we see merit in supporting that use case in Mockito, we will keep this issue open.

**Comment by raphw** [comment]
I just had a look at the suggestion of @ChristianSchwarz - nicely done! (Could you squash the commits and remove the formatting/import changes?)

I will try to make this my summer project. I know a lot of people want this and I just had too much Byte Buddy maintenance lately to commit to it but I still really want to do it. I have to have a deeper look at Christian's solution and check all edge cases but I think the API makes sense and after some clean up, this might be the way to go!

**Comment by raphw** [comment]
Alright, something bit me yesterday and I built a POC with a slightly different approach. What always bothered me about static mocks was their potential inferrence with parallel test cases and so forth, but all of this is solvable. I put the working prototype on a branch and suggest this API:

https://github.com/mockito/mockito/blob/static-mock/subprojects/inline/src/test/java/org/mockitoinline/StaticMockTest.java

**Comment by CharlieReitzel** [comment]
Hi Rafael,  I like the scoping of the mock via Closable interface.   +1.   Also, Java 8 method references make a lot of sense in this context.  

Just out of interest, this looks a bit different than Christian's use of lambdas.  Or does it work out the the same thing (w/ the Java compiler handing Mockito a `Function`  to mock, either way)?

**Comment by TimvdLippe** [comment]
Thanks @raphw  for that and glad to see some progress on it. I do want to reiterate that you should not take this project on if you do not have the mental/physical capacity to do so. As a fellow maintainer, I would prefer if you stay in the running, rather than overloading you with this project.

Let us know if we can assist you in any matter possible. Very glad to see an elegant solution this problem, so props for figuring this out! Evidence yet again we can be lucky we have you on our side working on these problems 😄

**Comment by ChristianSchwarz** [comment]
@raphw glad you took some inspiration and motivation out of my attempt. I had quick look into  your branch, looks like you are going full steam ahead. 

Some questions and toughts:

-  The scope meant to automatically reset the static mock after usage, right?
   - It is safe when used with "try with resource" but people might use it without, its not possible to enforce it.
   - The try with resource adds noise / ceremony to the test it self , an other option is to do the reset automatically after a test.
   - How to mock multiple static classes?
   - Can I stub a static method in a before-each so it is not required to stub it in every test?

 - On twitter you wrote the static mocks are thread local
  - What if I am testing a larger component that runs code in multiple threads, how to do that?

**Comment by raphw** [comment]
Hi Christian,

thanks for your prototype. It helped me to put the last bits of the puzzle on its place. The thing that bothers me the most with static method mocks is the inability to run tests in parallel. It can really add a lot of confusion since not everybody expects it, for example when adjusting a build script. This way, it is fully safe. I think for now, we leave it this way; it would however not be difficult to add a `MockSettings` option to include additional threads or to even make the mock global. I still think we leave it out of the first version and consider it for later.

Mocking multiple classes for static methods is fully possible:

```
try (
  MockedStatic<Foo> foo = Mockito.mockStatic(Foo.class);
  MockedStatic<Bar> bar = Mockito.mockStatic(Bar.class)
) {
  // ...
}
```

The only thing that is not allowed is to mock the same class twice but Mockito will detect this and throw an exception. I also adjusted the JUnit integration to make the ceremony superfluous. You can simply do:

```java
@Mock 
MockedStatic<Foo> foo;
```

in a JUnit test when using the rule or extension and the same works in JUnit 5's extension for method parameters. Mockito will then take care of the lifecycle. If the explicit model is used and the mock is not closed, it is currently unsafe. We could however add functionality to clear all mocks on a thread within the runner and rules where I do however expect the annotations to be used mostly.

**Comment by ChristianSchwarz** [comment]
@raphw I know I am a little late to the party. Is it possible to simplify the API a bit so users are not forced to use the MockStatic instance like this:


```
try(MockedStatic<Dummy> ignore = Mockito.mockStatic(Dummy.class)) {
            when(Dummy::foo).thenReturn("bar");
            verify(Dummy::foo);
            
}
```

Once the `Verification`-Lambda is executed you know that `Dummy.foo()` was called. Then you can map the `Dummy.class` to the corresponding `MockedStatic` object. This way users don't need to reference the `MockedStatic` object. This also allows to simplify futher by obmitting the try-with-resource. E.g:

```
Mockito.mockStatic(Dummy.class, ()=>{  //Note: The lambda defines the scope of the static mock
  when(Dummy::foo).thenReturn("bar");
  verify(Dummy::foo);
});
```

**Comment by TimvdLippe** [comment]
I don't think I would support a lambda-solution or a solution that relies on a return type. If we would return a special interaction method, where you later have to call `close()` yourself, then you can assign this to any variable/field and make it longlived. This opens up a whole can of worms with working with static method mocking.

Therefore, for me, one requirement for static method mocking is that it is explicitly limited to scope and can't escape that scope. The `try`-solution addresses this point, but the lambda return call does not.

**Comment by ChristianSchwarz** [comment]
When you use the lambda you will be safe, the static mock will only life in the scope of the lambda.

```
Mockito.mockStatic(Dummy.class).scope(() -> { 
  ... 
});
```
The mock is generated when you call the `scope(..)` method, then the lambda/scope is executed.After that the mock will be closed/reset. There is no way to escape the scope. But this extra method and lamda add complexity, same goes for the overloaded mockStatic method that accepts the scope/lambda. 

----

The `MockedStatic` approach don't give you guarantees if the user doesn't use try-with resource like this:

```
var dummy= Mockito.mockStatic(Dummy.class);
dummy.when(Dummy.foo()).thenXXX();
```
The mocking/stubbing/verifing will work as expected but leaves the Class in a mocked state. 


The automatic close/reset at the end of a test is an option that can be considered.

**Comment by mkq** [comment]
Replacing `Mockito.mockStatic(klass, () -> { … })` with the fluent `Mockito.mockStatic(klass).scope(() -> { … })` may look nice (to others, not necessarily to me :-)), but then the compiler won't complain when the scope is missing.
If you want to force a scope, please make it a parameter of (all overloaded) `mockStatic`.

**Comment by raphw** [comment]
The problem with only allowing it in a given lambda scope is that any alternartive life-cyle management approaches are not supported. Having the construct available for example allows for the use in Junit extensions and rules.

The try-with-resources construct can be forgotten but it is also the most flexible option. I suggest we stick to it for the first release.

**Comment by kriegfrj** [comment]
I'm late to this conversation, and I'm glad to see that there is some movement on this. I am firmly in support of the addition of static method mocking to Mockito. It is true that there are workarounds - someone wrote: "Every problem can be solved with a layer of abstraction" - but then there is the corollary that seems to be forgotten: "...except for the problem of too many layers of abstraction." Simplicity should be a goal of code - we shouldn't add complexity without good reason, and layers of abstraction add complexity at all levels (in development and at runtime). In the case of mocking statics, it is certainly *possible* to get around this problem (to an extent) by adding a new layer of abstraction - but this adds complexity. We may often decide the trade-off is worth it to make our code testable, but wouldn't it be better if you didn't have to make the trade-off? If you could mock statics without paying the price of adding an extra level of indirection? Also, adding a layer of abstraction is only possible in code that I have control over. What if I'm using a third-party library? `System.currentTimeMillis()` is a classic example - there's no good way to reliably simulate code running at different system times without mocking this static method (you can use wrappers in your own code, but there's no guarantee that all the 3rd-party libraries you might want to use will).

I also wanted to throw a couple of things into the mix, in commentary on the current implementation-in-progress:

1. _constructors are effectively static methods_. I think it would be great if whatever mechanism is implemented here to mock static methods can also be used to mock constructors. Doing so gives you a way to inject mock instances of dependencies into your class-under-test without requiring you to rewrite the class-under-test for this purpose. JMockit provides such a capability, and I would love to see Mockito have it too (as JMockit doesn't play will with OSGi, whereas Mockito does).
2. There has been talk about the temporal scope of the static mock, but I think that consideration also needs to be given to the caller scope: which classes will see the mocked method and which will see the original. I'm thinking here in particular the case of mocking methods of common classes (eg, JDK classes) that might also be in use by the test framework. The test framework will rely on the normal behaviour of the mocked class while you will want the class-under-test (and its dependencies) to see the mocked behaviour. It wouldn't be a good idea to mock (eg) `System.currentTimeMillis()` if it was going to be seen by the test framework as it will mess up the running of the test, but it would be great to be able to mock it for the code-under-test.

**Comment by raphw** [comment]
Thanks, this is about my train of thought. Additionally, it is sometimes just not feasible to refactor code. Either the code is not yours or you cannot change a legacy code base (immediately). In these cases, it is very convenient to have the possibility to mock static methods, also to give you safety if you are changing code in intermediate steps to finally remove them. 

Construction mocks are in the making and will be part of 3.5.0. You can have a look at my open PRs if you want to give it a test run. It's fully working and I am only waiting for the Mockito team members to return from their vacations to get some feedback from them before merging.
